code/modules/recycling/compactor.dm

```diff
@@ -0,0 +1,139 @@
+		var/obj/item/trashcube/T =  new /obj/item/trashcube(get_turf(src))
+		for(var/obj/item/O in src)
+			flush_count = max(0, flush_count-3) //10% charge drained for each item
+			if(istype(O,/obj/item/trashcube) || O.w_class > W_CLASS_LARGE)
+				visible_message("<span class='warning'>\The [src] groans and spits out \the [O], prematurely ending the cycle.</span>")
+				launch(O)
+				flick("compactor_jobcancel",src)
+				break
+			O.forceMove(T)
+			T.w_class = max(T.w_class,O.w_class-1) //Make a cube starting at SMALL size, but increase it to medium if there is a large item inside.
+			T.update_icon()
+		if(emagged)
+			for(var/mob/M in src)
+				var/obj/item/weapon/reagent_containers/food/snacks/monkeycube/humancube/H = new(src)
+				H.contained_mob = M
+				M.forceMove(H)
+				launch(H)
+		if(!T.contents.len)
+			qdel(T) //If somehow we ended up with nothing inside
+
+/obj/machinery/disposal/compactor/proc/launch(var/atom/movable/AM)
+	AM.forceMove(get_turf(src))
+	var/turf/target = get_offset_target_turf(src.loc, rand(5)-rand(5), rand(5)-rand(5))
+	AM.throw_at(target, 5, 1)
+
+/obj/machinery/disposal/compactor/handle_trunk()
+	return
+
+/obj/machinery/disposal/compactor/update_icon()
+	icon_state = "compactor_[stat & NOPOWER ? "off" : "on"]"
+
+/obj/machinery/disposal/compactor/Topic(href, href_list)
+	if(usr.loc == src)
+		to_chat(usr, "<span class='warning'>You cannot reach the controls from inside.</span>")
+		return
+	if(!anchored)
+		return 1
+	if(..())
+		usr << browse(null, "window=disposal")
+		usr.unset_machine()
+		return 1
+	else
+		src.add_fingerprint(usr)
+		usr.set_machine(src)
+
+		if(href_list["compact"])
+			if(flush_count < flush_every_ticks || flush || !contents.len) //Not charged OR current working OR nothing inside
+				return
+			if(!allowed(usr) && !emagged) //Currently can't emag it anyway
+				to_chat(usr, "<span class='warning'>Access denied.</span>")
+				return
+			for(var/mob/M in src)
+				if(!emagged) //Currently cannot be emagged, add code here if desired
+					visible_message("<span class='warning'>The safety light flashes on \the [src].</span>")
+					flick("compactor_error",src)
+					return
+			compact()
+			update_icon()
+		if(href_list["eject"])
+			eject()
+
+		nanomanager.update_uis(src)
+
+/obj/machinery/disposal/compactor/process()
+	updateDialog()
+	update_icon()
+	if(stat & NOPOWER || stat & BROKEN)
+		return
+	if(!anchored)
+		return
+	//No idle power usage, unlike a normal disposal.
+	if(flush_count < flush_every_ticks) //Compactors don't autocompact, but they do need to charge up over 30 ticks. We'll repurpose those variables here.
+		use_power(500)
+		flush_count++
+
+/obj/machinery/disposal/compactor/attackby(var/obj/item/I, var/mob/user)
+	add_fingerprint(user)
+	if(iswrench(I)) //We want this to be a high level operation, before any of the place in bin code or disassemble bin code
+		wrenchAnchor(user, 30)
+		power_change()
+		return
+	if(!emagged && istype(I,/obj/item/weapon/card/emag))
```

Comment:
There's an `EMAGGABLE` flag. Hardcoding it into attackby is not the way to do it.